apps/labs/posts/integrating-hypothesis-into-sympy.md

```diff
@@ -0,0 +1,97 @@
+
+There are many ways to test your software : unit testing, regression testing, diff testing, and mutation testing are a few that come to mind. This blog post is for anyone interested in understanding the value of utilizing _property based testing_ in their software projects. The post  will be broken up into three parts:
+
+* What is Property Based Testing?
+* What is Hypothesis?
+* Experience Integrating into SymPy
+
+If you wish to follow the examples in the blog post, you will need to have the latest version of sympy (on master) and you need to install Hypothesis via:
+
+```bash
+pip install hypothesis
+```
+# What is Property Based Testing?
+
+Property based testing (PBT) is a technique where instead of testing individual test cases, you specify properties you which to hold true for a range of inputs. These properties are then tested against automatically generated test data. PBT uses logical properties over generated test data to facilitate broad, robust testing that can expose edge cases not easily found via traditional test cases.
+
+PBT _shines_ when testing generic functions and libraries working across a wide range of possible inputs, where manually enumerating test cases is difficult but the behavior in question is easily testable. Examples include:
+
+* Mathematical Operations
+* String formatting
+* Database migration
+* Compression algorithms
+
+Let's take a simple, concrete example. Say we want to ensure that after multiplying two polynomials together, the degree of the resulting polynomial is the sum of the degrees of the two polynomials. We could write a test case for this:
+
+```python
+from sympy.abc import x
+def test_degree():
+    f = Poly(x**2 + 1, x)
+    g = Poly(x**3 + 1, x)
+    h = f*g
+    assert h.degree() == f.degree() + g.degree()
+```
+Notice, we are limited in how many various combinations of _f_ and _g_ we can test. It would be better if we could fix a property and have a library automatically generate interesting test cases and run them for us. We would no longer need to worry about thinking up input/output pairs. This would increase our **_trust_** in the implementation.
+
+# What is Hypothesis?
+ > [Hypothesis](https://hypothesis.readthedocs.io/en/latest/) is a Python library for creating unit tests which are simpler to write and more powerful when run, finding edge cases in your code you wouldn’t have thought to look for. It is stable, powerful and easy to add to any existing test suite.
+> Now, let's test out the property above using Hypothesis:
+  
+  ```python
```

Comment:
I can't see the preview, but the indentation here may cause problems there. You probably want to have
```suggestion
```python
```